core/src/main/java/io/grpc/internal/DelayedStream.java

```diff
@@ -56,263 +54,228 @@
-      realStream.setDecompressor(decompressor);
-    }
-    if (compressor != null) {
-      realStream.setCompressor(compressor);
-    }
+  @Override
+  public void start(final ClientStreamListener listener) {
+    checkState(this.listener == null, "already started");
+    boolean startStream;
 
-    for (PendingMessage message : pendingMessages) {
-      realStream.setMessageCompression(message.shouldBeCompressed);
-      realStream.writeMessage(message.message);
-    }
-    // Set this again, incase no messages were sent.
-    realStream.setMessageCompression(messageCompressionEnabled);
-    pendingMessages.clear();
-    if (pendingHalfClose) {
-      realStream.halfClose();
-      pendingHalfClose = false;
-    }
-    if (pendingFlowControlRequests > 0) {
-      realStream.request(pendingFlowControlRequests);
-      pendingFlowControlRequests = 0;
+    Status savedError;
+    synchronized (this) {
+      this.listener = checkNotNull(listener, "listener");
+      savedError = error;
     }
-    if (pendingFlush) {
-      realStream.flush();
-      pendingFlush = false;
+    if (savedError != null) {
+      listener.closed(savedError, new Metadata());
+      return;
     }
-    // Ensures visibility.
-    startedRealStream = realStream;
-  }
 
-  /**
-   * Transfers all pending and future requests and mutations to the given stream.
-   *
-   * <p>No-op if either this method or {@link #cancel} have already been called.
-   */
-  final void setStream(ClientStream stream) {
-    synchronized (this) {
-      if (error != null || realStream != null) {
-        return;
-      }
-      realStream = checkNotNull(stream, "stream");
-      // listener can only be non-null if start has already been called.
-      if (listener != null) {
-        startStream();
+    delayOrExecute(new Runnable() {
+      @Override
+      public void run() {
+        realStream.start(listener);
       }
-    }
+    });
   }
 
   @Override
-  public void writeMessage(InputStream message) {
-    if (startedRealStream == null) {
-      synchronized (this) {
-        if (startedRealStream == null) {
-          pendingMessages.add(new PendingMessage(message, messageCompressionEnabled));
-          return;
+  public void writeMessage(final InputStream message) {
+    checkNotNull(message, "message");
+    if (passThrough) {
+      realStream.writeMessage(message);
+    } else {
+      delayOrExecute(new Runnable() {
+        @Override
+        public void run() {
+          realStream.writeMessage(message);
         }
-      }
+      });
     }
-    startedRealStream.writeMessage(message);
   }
 
   @Override
   public void flush() {
-    if (startedRealStream == null) {
-      synchronized (this) {
-        if (startedRealStream == null) {
-          pendingFlush = true;
-          return;
+    if (passThrough) {
+      realStream.flush();
+    } else {
+      delayOrExecute(new Runnable() {
+        @Override
+        public void run() {
+          realStream.flush();
```

Comment:
I was okay with replaying each flush. I wasn't too concerned about that level of performance in the slow case. It should be possible to coalesce flushes, but is slightly non-trivial because of interactions with other methods like `halfClose()` and `cancel()`. I'm don't have much problem calling optimizing flush as future work, but I do have some trouble caring about the case at present because it will be _no worse_ than when `DelayedStream` is not being used.